### PR TITLE
Support Blackfly USB3 camera

### LINKED ISSUE #2: Support Blackfly USB3 camera
Implement the spinnaker drivers such that basic usage of a Blackfly USB3 camera is supported.

Specifically we are targeting the [BFLY-U3-23S6M-C model](https://www.flir.com/products/blackfly-usb3/?model=BFLY-U3-23S6M-C&vertical=machine+vision&segment=iis).

Basic usage is defined as the supported behaviors implied by [the tests in the acquire-driver-egrabber repo](https://github.com/acquire-project/acquire-driver-egrabber/tree/d9724542a8fadb053fa03d1cfc17e637a95fec04/tests).

Similarly, the implementation could and probably should mirror acquire-driver-egrabber since both SDKs rely on GenICam definitions under the hood.

### PR DESCRIPTION
**Pull request by andy-sweet**
This uses https://github.com/acquire-project/acquire-driver-egrabber as a template and modifies the driver implementation in spinnaker.cpp to support the Spinnaker SDK.

Some implementation details are specific to the Blackfly camera and will need to be generalized when targeting other cameras.

Closes #2

### CODE REVIEW

**Comment by andy-sweet** [comment]
Marked as draft because this probably shouldn't be merged yet, but it is generally ready to review.

I've built this and run tests locally on macOS and Windows.

The main file to focus on is `spinnaker.cpp`, as the other files were mostly unchanged from the egrabber repo. Some focus areas in the review I'd appreciate.

- Should we make the readable/writable checks more consistent for getting/setting spinnaker/GenICam fields?
- Should we always error when trying to set a non-settable field?
- Better ideas for Acquire <-> Spinnaker enum translation (compared to the switch-case converter functions I defined).
- Check out the remaining TODOs.
- Should we expect the Spinnaker SDK to be installed on the testing system? Do we need to package/distribute it here? Can we?
- Need to add GitHub actions with a runner that has a Blackfly camera connected.

File: LICENSE
Comment (aliddell) [inline LICENSE:189]: Why change this back?

Comment (andy-sweet) [reply LICENSE:189]: If you read the text above the copyright notice, I think this is intended as a template. This also explains why even Apache projects leave it like that (e.g. [kafka](https://github.com/apache/kafka/blob/trunk/LICENSE)).

Based on the same text, I think actual copyright notice referencing the owner/licensor should appear in each source code file or maybe even every file that's distributed.

Also, I'd prefer to do all the licenses at once and consistently, rather than make this a special case.

**Comment by andy-sweet** [comment]
Back to draft while I address comments from the live review.

**Comment by andy-sweet** [comment]
There are still a few TODOs and I want to add some more test coverage around setting properties, but this is generally ready for review again. I made a few commits since actually testing this with cameras, but I think those should be safe.

I am still having issues where spinnaker complains about a lack of memory when initializing the camera. I found [a way to decrease the probability of this happening](https://github.com/acquire-project/acquire-driver-spinnaker/pull/4/files#r1303325868) but it was not a complete fix.

File: src/spinnaker.cpp
Comment (andy-sweet) [inline src/spinnaker.cpp:414]: Sometimes spinnaker can throw here, complaining that the device does not have enough memory for buffer allocation. Changing the buffer handling mode to newest only seemed to help, but did not entirely fix things.

```cpp
    Spinnaker::GenApi::INodeMap& stream_node_map = camera->GetTLStreamNodeMap();
    Spinnaker::GenApi::CEnumerationPtr stream_handling_node = stream_node_map.GetNode("StreamBufferHandlingMode");
    *stream_handling_node = "NewestOnly";
```

File: src/spinnaker.cpp
Comment (andy-sweet) [inline src/spinnaker.cpp:149]: I've flip-flopped between guarding with an `if` and just `throwing`. I think the semantics of `set` mean that we want to try to set as many new parameter values as we can even if one fails, so I'm using the `if` guard right now. Maybe I should just wrap the individual setters in try/catch blocks and let spinnaker throw though?

File: src/acquire-core-libs
Comment (nclack) [inline src/acquire-core-libs:1]: looks like this might be pointing to your fork rather than [f56ba33](https://github.com/acquire-project/acquire-core-libs/commit/f56ba33ebfd5e6971ad8b7b502d76fb013de7d1e)

Comment (andy-sweet) [reply src/acquire-core-libs:1]: Fixed! I think... I just forgot to pull my core-lib's main and push the reference up on this branch.

File: .github/workflows/test_pr.yml
Comment (andy-sweet) [inline .github/workflows/test_pr.yml:6]: Reminder for me.

```suggestion
  pull_request_target:
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]: Why not use `std::clamp`? Is it to avoid including `<algorithm>`?

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]: Why is this here twice?

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_exposure_time_capabilities(
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_binning_capabilities(CameraPropertyMetadata* meta) const
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_roi_offset_capabilities(
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_roi_shape_capabilities(CameraPropertyMetadata* meta) const
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_pixel_type_capabilities(
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]:
```suggestion
SpinnakerCamera::query_triggering_capabilities(CameraPropertyMetadata* meta)
```

File: src/spinnaker.cpp
Comment (aliddell) [inline src/spinnaker.cpp]: When do you think you'll revisit this?

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: No. I didn't know `clamp` was brought into `std`. Changed!

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Great question. I don't know. I swear at some point I was comparing two different things here.

Given the current hardcoding of triggers for the blackfly camera in this PR, I think this should probably be checking that the LineSelector is line 1, so I made that change. This will likely change quite a bit after adding support for the oryx camera.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: I'm not sure, partially because I'm not sure where this logic should live - here, in SpinnakerCamera's constructor, or in the acquire video runtime.

In general, the remaining TODOs are ones that I could not resolve on my own, so any input would be much appreciated.

Would be great to have no remaining TODOs before merging. Or if there are some, they should be tracked by GitHub issues.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Oops. Thanks for the catch. I fixed these locally so that both the declarations and definitions were fixed in one batch.

**Review by andy-sweet — COMMENTED** [review]
Thanks for the review!

File: src/spinnaker.cpp
Comment (aliddell) [reply src/spinnaker.cpp]: This is roughly analogous to the camera restart logic we (mostly Nathan) wrote into `dcam_start`. I think it makes sense to live here, but unless you're actually seeing the problem (and / or can reproduce it in a test) I would say it's fine to leave it for now as long as you make an issue.

**Review by aliddell — APPROVED** [review]

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: better to use `frame->GetFrameID()` if you can

File: src/spinnaker.cpp
Comment (nclack) [reply src/spinnaker.cpp]: > I'm not sure where this logic should live

The logic should live here. I suspect [`DeInit()`](http://softwareservices.flir.com/Spinnaker/latest/class_spinnaker_1_1_camera.html#ae683d1e0dd5b7166d03468c06e04907c) gets called when `Camera` destructor goes. But if you're worried about it I don't see a reason not to handle it here.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: I commented on this below but `GetFrameID` is preferred for the hardware frame id. We want whatever is closest to the camera hardware.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp:1018]: no change needed. But fwiw `auto` is ok.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: Might be nice to know the frame id here.

```suggestion
        LOGE(
          "Image incomplete: (%d) %s",
          (int)frame->GetFrameID(),
          Spinnaker::Image::GetImageStatusDescription(frame->GetImageStatus()));
```

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: I believe there are two ways. I like the second the most.

1. Set the LineSelector to Line0, since that line doesn't support output.
2. Set the LineSource to UserOutput1. Then based on whether you want the line to be high or low at rest, set the UserOutputValue1. It might be best to leave that undefined.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: I believe this is based on the `LineInverter` prop. If true, the level is low when active.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp:564]: ❤️

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: I'd prefer not erroring. should log and read back what the camera is set to.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: `last_known_settings_` might be used in two ways. First, to quickly return the last known settings via get. Second, as a memoizing strategy to reduce the number of Spinnaker calls make on `set`.

I think that Second use case is the important one. And so that would be why you might want to be sure to update it here.

File: src/spinnaker.cpp
Comment (nclack) [inline src/spinnaker.cpp]: There's an order of operations thing here that I probably also have wrong in egrabber.

Depending on how the camera works you may need to set offset to (0,0) before changing the shape so you have the full range of the sensor. Then set the offset to the desired one.

File: src/spinnaker.cpp
Comment (nclack) [reply src/spinnaker.cpp:414]: It might be worth being opinionated about the stream handling so we get consistent behavior in the field.

File: cmake/git-versioning.cmake
Comment (nclack) [inline cmake/git-versioning.cmake]: We get a lot of poop in the build logs because this is a global `add_definitions'. Maybe we can start doing better with something like this suggestion and then use the `target_add_git_versioning()` function to the main driver target. 

```suggestion
set(CPACK_PACKAGE_VERSION ${GIT_TAG})
message(STATUS "Version ${GIT_TAG} ${GIT_HASH}")
function(target_add_git_versioning tgt)
    target_compile_definitions(${tgt} 
        -DGIT_TAG=${GIT_TAG}
        -DGIT_HASH=${GIT_HASH}
    )
endfunction()
```

Comment (nclack) [reply cmake/git-versioning.cmake]: If we figure it out here we can copy the pattern around.

File: src/CMakeLists.txt
Comment (nclack) [inline src/CMakeLists.txt:27]: See my comment in `git_versioning.cmake`

```suggestion
target_add_git_versioning(${tgt})
set_target_properties(${tgt} PROPERTIES
        MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>"
        )
```

**Review by nclack — CHANGES_REQUESTED** [review]
Tried to answer questions and had minor suggestions. Marking as request changes just so we don't merge before you've had a chance to look.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Done.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Done.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp:1018]: With Spinnaker, I prefer being more explicit about the types because I was initially a bit confused, so going to leave this.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: The problem with both uses is that we're assuming that Acquire is the only thing modifying the camera settings while its process is alive. That might be true in most cases, but it's not guaranteed to be true (e.g. I think I can successfully modify settings with SpinView while Acquire is running).

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: I ran into a case like that hence the comment above. But I think you're right that I may be not handling all cases. Will ensure that tests cover those.

Do we assume that shape takes precedence over offset? E.g. what if a user has previously set a non-zero offset, then sets the shape to the same size as the sensor? In that case, I think we should refuse to set the shape and at least log an error before continuing with other properties.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Done.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Done.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: I like (2), so I did that. I only set UserOutputValue1 to false/off when this trigger is disabled. Don't have access to the camera right now, but I'm going to push and see what happens.

Comment (andy-sweet) [reply src/spinnaker.cpp]: Not working yet. I'll debug this tomorrow.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Not working yet. I'll debug this tomorrow.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: For now, I've removed the TODO and left in the read at the end of each setter.

File: src/spinnaker.cpp
Comment (andy-sweet) [reply src/spinnaker.cpp]: Moved it from here to the constructor. That didn't seem to cause any new problems, but I'll verify tomorrow.

File: cmake/git-versioning.cmake
Comment (andy-sweet) [reply cmake/git-versioning.cmake]: Agreed on the poop. If that is also in egrabber (and maybe others), I'd prefer to create an issue to follow up rather than work out all the details in this PR.

Comment (andy-sweet) [reply cmake/git-versioning.cmake]: Ah wait, now I saw the sibling comment. I effectively made those changes and the poop is gone now. How can I verify that GIT_TAG and GIT_HASH are being defined/used as intended?

Comment (andy-sweet) [reply cmake/git-versioning.cmake]: I think I verified by grepping the build directory for those two and found that the compile commands use the correct GIT_HASH for this source, while leaving the others alone to use theirs (without the redefine). So I'm going to push this.

File: src/spinnaker.cpp
Comment (nclack) [reply src/spinnaker.cpp]: It's a bit like we're setting both at the same time from the perspective of the Acquire api, but spinnaker only allows changing one at a time.

The problem I've run into is when they have set a valid roi with a non-zero offset previously and then want to return to full sensor. The offset (to 0) needs to be applied before the shape change.

On the other hand, if I'm going in the other direction (full sensor to an valid offset roi), the new shape has to be applied before changing the offset.

In both cases, the user has valid settings. It's just that we need to watch the order of operations so things don't fail.

Comment (andy-sweet) [reply src/spinnaker.cpp]: That's clear.

But what if they don't have valid settings? I assume we should not change either offset or shape if the combination is invalid? I.e. given Acquire treats these two as coupled the change should happen atomically or not at all (rather than with offset changed, but not shape).

Comment (andy-sweet) [reply src/spinnaker.cpp]: Discussed in person and decided to only validate the region in this PR. If it's not valid, we don't try to set it.

**Comment by andy-sweet** [comment]
@nclack : this is ready for another pass. I added a test for the ROI validation logic and implemented some logic to handle that validation. I also got some of the more advanced output line behavior implemented and tested too.

**Review by nclack — APPROVED** [review]
LGTM!

**Pull request merged**
